onnx/defs/nn/defs.cc

```diff
@@ -2190,6 +2190,193 @@ ONNX_OPERATOR_SET_SCHEMA(
+    N = input_shape.dim(0); // Otherwise, N is unknown.
+  }
+  *final_image_shape->add_dim() = N;
+
+  if (block_shape_size > 0) {
+    C = input_shape.dim(1) / block_shape_size; // Otherwise, C is unknown.
+  }
+  *final_image_shape->add_dim() = C;
+
+  // Image dimensions are dynamic
+  for (size_t i = 0; i < n_input_dims; ++i) {
+    Dim image_dim_i;
+    if (image_shape.size() > 0) {
+      image_dim_i.set_dim_value(image_shape[i]); // Otherwise, spatial dimensions are unknown
+    }
+    *final_image_shape->add_dim() = image_dim_i;
+  }
+  return;
+}
+
+static const char* Col2Im_ver17_doc = R"DOC(
+The operator rearranges column blocks back into a multidimensional image
+
+Col2Im behaves similarly to PyTorch's fold https://pytorch.org/docs/stable/generated/torch.nn.Fold.html,
+but it only supports *batched* multi-dimensional image tensors.
+
+NOTE: Although specifying image_shape looks redundant because it could be calculated from
+      convolution formulas, it is required as input for more advanced scenarios as explained
+      at PyTorch's implementation (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Col2Im.cpp#L10)
+
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Col2Im,
+    17,
+    OpSchema()
+        .Attr(
+            "dilations",
+            "1-dimensional tensor with dilation value along each spatial axis of the image. "
+            "If not present, the dilation defaults to 1 along each spatial axis of the image.",
+            AttributeProto::INTS,
+            OPTIONAL_VALUE)
+        .Attr(
+            "pads",
+            "1-dimensional tensor with padding value for the beginning and ending along each spatial axis, "
+            "it can take any value greater than or equal to 0. "
+            "The value represent the number of pixels added to the beginning "
+            "and end part of the corresponding axis. `pads` format should be as follow "
+            "[x1_begin, x2_begin...x1_end, x2_end,...], where xi_begin is the number of pixels "
+            "added at the beginning of axis `i` and xi_end is the number of pixels added at the end of axis `i`. "
+            "If not present, the padding defaults to 0 along start and end of each spatial axis.",
+            AttributeProto::INTS,
+            OPTIONAL_VALUE)
+        .Attr(
+            "strides",
+            "1-dimensional tensor with stride value along each spatial axis. "
+            "If not present, the stride defaults to 1 along each spatial axis.",
+            AttributeProto::INTS,
+            OPTIONAL_VALUE)
+        .SetDoc(Col2Im_ver17_doc)
+        .Input(
+            0,
+            "input",
+            "Input data tensor to be rearranged from column blocks back into an image."
+            " This is a 3-dimensional tensor containing [N, C * n-ary-product(block_shape), L],"
+            " where N is batch dimension, C is image channel dimension and L is number of blocks.",
```

Comment:
For the middle-dimension, I believe that the order is block-major, correct? Also, is it correct to say that `n-ary-product(blockshape)` is the `dilated-block-size`? If so, I suggest replacing it with `dilated-block-size`.

Reply:
It will also help to specify the order of the blocks. Can we say something like
```
The blocks are enumerated in increasing lexicographic-order of their indices.
For example, with an image-size 10x20 and block-size 9x18, we have 2x3 blocks,
enumerated in the order block(0,0), block(0,1), block(0,2), block(1,0), block(1,1), block(1,2).
```

Reply:
> For the middle-dimension, I believe that the order is block-major, correct? Also, is it correct to say that `n-ary-product(blockshape)` is the `dilated-block-size`? If so, I suggest replacing it with `dilated-block-size`.

I think this was an implementation detail for reusing ORT's math::Col2ImNd that relied on ORT's tweaked math::Im2Col(accumulate=true). In that case, the original image size and block size had to be adjusted before calling math::Col2ImNd, which in turn, called im2Col. However, from the user point of view, they can specify only the original image shape and block shape and the implementor would be able to derive the dilated versions, if their implementation requires them.